internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,123 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) string
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	url          string
+	clientId     string
+	clientSecret string
+	token        string
+	offlineToken string
+	expiration   time.Time
+	mutex        sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	TokenType   string `json:"token_type"`
+}
+
+// NewLazyToken creates a new LazyToken instance.
+func NewLazyToken(url, clientId, clientSecret string) *LazyToken {
+	return &LazyToken{
+		url:          url,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+	}
+}
+
+// Token returns a valid token, acquiring a new one if necessary.
+func (lt *LazyToken) Token(ctx context.Context) string {
+	//echoCtx, ok := ctx.(echo.Context)
+	//	if !ok {
+	//	// Handle the case where the context is not an echo.Context
+	//	return ""
+	//	}
+	lt.mutex.Lock()
+	defer lt.mutex.Unlock()
+
+	if lt.token == "" || lt.expiration.Sub(time.Now()) < time.Minute {
+		//	echoCtx.Logger().Info("Acquiring new token")
+
+		// Acquire a new token and set the expiration time
+		lt.token = AcquireToken(lt.url, lt.clientId, lt.clientSecret, lt.offlineToken)
+		lt.expiration = time.Now().Add(tokenExpirationDuration)
+	} else {
+		//	echoCtx.Logger().Debug("Token requested: %s (expires at %s)\n", lt.token, lt.expiration)
+	}
+
+	return lt.token
+}
+
+// AcquireToken is a placeholder function that simulates token acquisition.
+func AcquireToken(tokenURL, clientID, clientSecret, offlineToken string) string {
```

Comment:
This function must be not exported, the `Tokener` interface is supposed to be the only way how to work with this package.

```suggestion
func acquireToken(tokenURL, clientID, clientSecret, offlineToken string) string {
```